### ISSUE TITLE
Transitive dependency list

### ISSUE THREAD

**Issue by theBradfo** [helpscout/beacon-android-sdk-sample#66]
Beacon version: 1.0.8

Hello!  My company is considering integration with the beacon sdk for our android client, however we are also consciously and actively trying to reduce or limit our deliverable size.  The `beacon-ui` and `beacon-core` dependencies bring along a relatively lengthy list of transitive dependencies.  While i understand R8 will reduce the source included in these aar's, I can only assume much/most of these libraries are used internally by the beacon sdks, thus they wouldn't be eligible for removal at build time.  

Would it be possible to further modularize these sdk's or rethink the inclusion of some of these libraries used internally?  I can see a list of libraries included transitively that we as a client would never use (i.e. this includes `koin`, however we have a heavy usage on `dagger2`).

**Output from Gradle dependencies task**
```
+--- com.helpscout:beacon-core:1.0.8
|    +--- org.jetbrains.kotlin:kotlin-stdlib-jdk7:1.3.31 -> 1.3.50 (*)
|    +--- org.jetbrains.kotlinx:kotlinx-coroutines-core:1.0.0 -> 1.1.1 (*)
|    +--- org.jetbrains.kotlinx:kotlinx-coroutines-android:1.0.0 -> 1.1.1 (*)
|    +--- com.squareup.okhttp3:okhttp:3.9.0 -> 3.12.0 (*)
|    +--- com.squareup.retrofit2:retrofit:2.3.0 -> 2.5.0 (*)
|    +--- com.squareup.retrofit2:converter-moshi:2.3.0
|    |    +--- com.squareup.retrofit2:retrofit:2.3.0 -> 2.5.0 (*)
|    |    \--- com.squareup.moshi:moshi:1.4.0
|    |         \--- com.squareup.okio:okio:1.11.0 -> 1.15.0
|    +--- ru.gildor.coroutines:kotlin-coroutines-retrofit:1.0.0
|    |    +--- org.jetbrains.kotlin:kotlin-stdlib:1.3.11 -> 1.3.50 (*)
|    |    +--- org.jetbrains.kotlinx:kotlinx-coroutines-core:1.0.1 -> 1.1.1 (*)
|    |    \--- com.squareup.retrofit2:retrofit:2.4.0 -> 2.5.0 (*)
|    +--- com.squareup.okhttp3:logging-interceptor:3.9.0 -> 3.9.1 (*)
|    +--- com.facebook.stetho:stetho:1.5.0
|    |    +--- commons-cli:commons-cli:1.2
|    |    \--- com.google.code.findbugs:jsr305:2.0.1
|    +--- com.facebook.stetho:stetho-okhttp3:1.5.0
|    |    +--- com.facebook.stetho:stetho:1.5.0 (*)
|    |    +--- com.google.code.findbugs:jsr305:2.0.1
|    |    \--- com.squareup.okhttp3:okhttp:3.4.2 -> 3.12.0 (*)
|    +--- com.jakewharton.timber:timber:4.7.1 (*)
|    +--- com.jakewharton.threetenabp:threetenabp:1.1.1
|    |    \--- org.threeten:threetenbp:1.3.7
|    +--- org.jetbrains.kotlin:kotlin-android-extensions-runtime:1.3.31 -> 1.3.50 (*)
|    \--- androidx.core:core:1.0.0 -> 1.1.0-rc02 (*)
+--- com.helpscout:beacon-ui:1.0.8
|    +--- com.squareup.okhttp3:okhttp:3.9.0 -> 3.12.0 (*)
|    +--- com.squareup.retrofit2:retrofit:2.3.0 -> 2.5.0 (*)
|    +--- com.squareup.retrofit2:converter-moshi:2.3.0 (*)
|    +--- ru.gildor.coroutines:kotlin-coroutines-retrofit:1.0.0 (*)
|    +--- com.jakewharton.threetenabp:threetenabp:1.1.1 (*)
|    +--- org.koin:koin-core:2.0.0-rc-2
|    |    \--- org.jetbrains.kotlin:kotlin-stdlib:1.3.21 -> 1.3.50 (*)
|    +--- org.koin:koin-android:2.0.0-rc-2
|    |    \--- org.koin:koin-core:2.0.0-rc-2 (*)
|    +--- org.koin:koin-java:2.0.0-rc-2
|    |    \--- org.koin:koin-core:2.0.0-rc-2 (*)
|    +--- org.koin:koin-android-scope:2.0.0-rc-2
|    |    +--- org.koin:koin-android:2.0.0-rc-2 (*)
|    |    +--- androidx.appcompat:appcompat:1.0.0 -> 1.1.0-rc01 (*)
|    |    \--- androidx.lifecycle:lifecycle-common:2.0.0 -> 2.1.0-rc01 (*)
|    +--- org.koin:koin-android-viewmodel:2.0.0-rc-2
|    |    +--- org.koin:koin-android-scope:2.0.0-rc-2 (*)
|    |    \--- androidx.lifecycle:lifecycle-extensions:2.0.0 (*)
|    +--- de.hdodenhof:circleimageview:2.1.0
|    +--- com.github.bumptech.glide:glide:4.9.0 (*)
|    +--- com.github.bumptech.glide:okhttp3-integration:4.9.0
|    |    +--- com.github.bumptech.glide:glide:4.9.0 (*)
|    |    +--- com.squareup.okhttp3:okhttp:3.9.1 -> 3.12.0 (*)
|    |    \--- androidx.annotation:annotation:1.0.0 -> 1.1.0
|    +--- net.yslibrary.keyboardvisibilityevent:keyboardvisibilityevent:2.1.0
|    +--- org.jetbrains.kotlin:kotlin-android-extensions-runtime:1.3.31 -> 1.3.50 (*)
|    +--- com.helpscout:beacon-core:1.0.8 (*)
|    +--- com.google.firebase:firebase-core:16.0.8 -> 17.1.0 (*)
|    +--- com.google.firebase:firebase-messaging:17.5.0 -> 20.0.0 (*)
|    +--- androidx.lifecycle:lifecycle-extensions:2.0.0 (*)
|    +--- androidx.appcompat:appcompat:1.0.0 -> 1.1.0-rc01 (*)
|    +--- androidx.cardview:cardview:1.0.0 (*)
|    +--- com.google.android.material:material:1.0.0 -> 1.1.0-alpha09 (*)
|    \--- androidx.browser:browser:1.0.0
|         +--- androidx.core:core:1.0.0 -> 1.1.0-rc02 (*)
|         +--- androidx.annotation:annotation:1.0.0 -> 1.1.0
|         +--- androidx.interpolator:interpolator:1.0.0 (*)
|         +--- androidx.collection:collection:1.0.0 -> 1.1.0 (*)
|         \--- androidx.legacy:legacy-support-core-ui:1.0.0 (*)

```

**Comment by scottyab** [comment]
Hey @theBradfo We are currently including the SDK dependencies as `api` with the thinking that host apps could take advantage of those dependencies. We also use R8 to further reduce the final .aar file size. However, if you're not using those dependencies that's of little help. 

Potentially us switching to using `implementation` with R8 could improve things?  I'll do some checking to see if this makes a significant saving on .apk file size in our sample apps. 

I'm curious to understand what kind of .apk file size increase you are seeing with beacon-ui dependancy included? and are you using code shrinker such as R8/ProGuard? If you'd like to email this instead of having here in github, please do email help@helpscout.com FOA Scott Alexander-Bown. Thanks

**Comment by theBradfo** [comment]
Thanks for the response @scottyab! The beacon sdk introduced some dependency conflicts with [aging] version of glide my app currently depends on versus what beacon transitively brings along.  I will have to circle back to this sdk integration in a few weeks as time permits by first bringing our app up-to-date with glide, then I will be able to offer a comparable before/after difference in apk size for the app in question.

**Comment by scottyab** [comment]
In a spike branch using a sample project we use for internal testing I made the switch from `api` to `implementation` and it only has 0.1 MB reduction in the resulting .apk file size. I was surprised it was so little. Please let us know when you have a before/after a difference in your apk size to compare. Thanks

**Comment by scottyab** [comment]
Hey @sipersso sorry to hear you are having a conflict with our use of Koin. When we introduced it I did some trials with a host app using Koin and thought it we made it compatible but maybe somethings changed since then. I'll create a new internal issue to investigate this. I'm keen to make sure solve this do you have a sample project that demonstrates the `KoinAppAlreadyStartedException`? If it's something you'd like to privately share please email to help@helpscout.com FOA Scott Alexander-Bown. 

We added Koin in 1.0.2 so switching to 1.0.1 would be temporary workaround while we look at making it compatible.

**Comment by scottyab** [comment]
@theBradfo just to confirm we have an internal issue created to switch from `api` to `implementation` with the primary goal of stopping the type of conflict you're seeing with Glide.

**Comment by sipersso** [comment]
@scottyab sent you a sample app that reproduces the issue. The error message is pretty clear though. I assume that you are using startKoin to start the global koin context and if you do, it wouldn't be possible to do this in the app. I guess this is a problem for you since if you can't use loadKoinModules unless Koin is started and if you startKoin yourself, then the consuming app will get an exception when loading it. 

It also seems like you are delaying the call to startKoin until the activity is shown. This means that the consuming app can't use loadKoinModules either...

**Comment by scottyab** [comment]
@theBradfo We've resolved the dependency visibility issue in 1.0.9, thanks for raising. 

@sipersso We have a separate internal issue for Koin scoping ANDSDK-227 and we have the work scheduled. Based on a conversation with the Koin author switching to use our own Koin Context should segregate the SDK from the host app's Global Koin Context. However, the beta release of real-time chat functionality is taking our immediate priority. Thanks again for raising this.
